src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt

```diff
@@ -65,26 +66,68 @@ class RsExtractFunctionConfig private constructor(
     var visibilityLevelPublic: Boolean = false,
     val parameters: List<Parameter>
 ) {
+
     private val parametersText: String
         get() = parameters.joinToString(",") { it.text }
+
     val argumentsText: String
         get() = parameters.filter { it.type != null }.joinToString(",") { it.name }
 
     val signature: String
-        get() {
-            var signature = "fn $name($parametersText)"
-            if (returnValue != null) {
-                signature += " -> ${returnValue.type.insertionSafeText}"
-            }
+        get() = buildString {
             if (visibilityLevelPublic) {
-                signature = "pub " + signature
+                append("pub ")
+            }
+            append("fn $name$typeParametersText($parametersText)")
+            if (returnValue != null) {
+                append(" -> ${returnValue.type.insertionSafeText}")
             }
+            append(whereClausesText)
             val stmts = elements.map { it.text }.toMutableList()
             if (returnValue?.expression != null) {
                 stmts.add(returnValue.expression)
             }
-            return signature + "{\n${stmts.joinToString(separator = "\n")}\n}"
+            append("{\n${stmts.joinToString(separator = "\n")}\n}")
+        }
+
+    private val typeParametersText: String
+        get() {
+            val typeParams = typeParameters()
+            if (typeParams.isEmpty()) return ""
+            return typeParams.joinToString(separator = ",", prefix = "<", postfix = ">") { it.text }
+        }
+
+    private val whereClausesText: String
+        get() {
+            val wherePredList = containingFunction.whereClause?.wherePredList ?: return ""
+            if (wherePredList.isEmpty()) return ""
+            val typeParams = typeParameters().map { it.declaredType }
+            if (typeParams.isEmpty()) return ""
+            val filtered = wherePredList.filter { typeParams.contains(it.typeReference?.type) }
+            if (filtered.isEmpty()) return ""
+            return filtered.joinToString(separator = ",", prefix = " where ") { it.text }
+        }
+
+    private fun typeParameters(): List<RsTypeParameter> {
+        val bounds = typeParameterBounds()
+        val paramAndReturnTypes = (parameters.mapNotNull { it.type } + listOf(returnValue).mapNotNull { it?.type })
```

Comment:
`listOf(returnValue).mapNotNull { it?.type }` can be simplified to `listOfNotNull(returnValue?.type)`